Custom main and global executors

al45tair opened the pull request:
Add write up of custom executor proposals. There are still some concerns surrounding integration with Dispatch that need working through.

al45tair committed:
Initial public draft of custom executors proposal.

al45tair committed:
Updates following pitch comments.

Use `any TaskExecutor` instead of `any Executor` for `Task.defaultExecutor`.

Rename `ExecutorJobKind` to `ExecutorJob.Kind`.

Add `EventableExecutor`; replace `SerialRunLoopExecutor` with `MainExecutor`,
then make `MainActor` and `PlatformMainExecutor` use the new protocol.

al45tair committed:
Add `ExecutorJob.allocator` and move allocation methods there.

The previous iteration would have required us to make calling the
allocation methods a fatal error, and would have meant executors had
to check explicitly for the `.task` job kind before using them.

Instead, if we add a new `LocalAllocator` type, we can have an
`allocator` property that is `nil` if the job doesn't support allocation,
and a valid `LocalAllocator` otherwise.

al45tair committed:
Updated to match current implementation.

Since the previous pitch, I've added `Clock`-based `enqueue` methods
and made a few other changes, notably to the way you actually set
custome executors for your program.

al45tair committed:
Remove EventableExecutor, further updates from pitch.

Remove `EventableExecutor`; we can come back to that later.

Update documentation for `currentExecutor` and add `preferredExecutor` and
`currentSchedulableExecutor`.

Move the clock-based enqueuing to a separate `SchedulableExecutor` protocol,
and provide an efficient way to get it.  This means we don't need the
`supportsScheduling` property.

Back `ClockTraits` with `UInt32`.

al45tair committed:
Use typed throws rather than `rethrows`.

We should use typed throws, which will make this work better for Embedded
Swift.

FranzBusch commented on proposals/nnnn-custom-main-and-global-executors.md:
```diff
@@ -0,0 +1,873 @@
+  ///   1. The custom executor associated with an `Actor` on which we are
+  ///      currently running, or
+  ///   2. The preferred executor for the currently executing `Task`, or
+  ///   3. The task executor for the current thread
```
I think this should be "for the current task" instead of "current thread"

al45tair replied:
I'm not sure there's a practical difference there; the task executors are tracked using thread local data.

al45tair replied:
It's this code here that does the tracking: https://github.com/swiftlang/swift/blob/5a6a14785075436b8bdb2c423cd47df4945d3ec8/stdlib/public/Concurrency/Actor.cpp#L114

FranzBusch commented on proposals/nnnn-custom-main-and-global-executors.md:
```diff
@@ -0,0 +1,873 @@
+  /// Executors can implement this method explicitly to avoid the use of
+  /// a potentially expensive runtime cast.
+  @available(SwiftStdlib 6.2, *)
+  var asSchedulable: AsSchedulable? { get }
```
```suggestion
  var asSchedulableExecutor: SchedulableExecutor? { get }
```

al45tair committed:
Updates after second SE pitch.

Removed clock conversion functions; we're going to add `enqueue` and
`run` functions instead.

Updated comments for various items.

Mention that we're planning to expose the built-in executor implementations
by name.

al45tair committed:
Add link to third pitch.

al45tair committed:
Further proposal updates on the basis of feedback received.

Allow `DefaultExecutorFactory` to be defined in the `@main` struct
as well as at top level.

Remove some text from the proposal that is out of date.

Don't expose the Dispatch or CF executors, since that implies that we
will have a permanent dependency on Dispatch or Foundation.

al45tair committed:
Remove CooperativeExecutor from the public API.

We can have something similar in a separate package; let's not share
this implementation outside of the runtime for now (it has a number
of gotchas and we'd have to spend quite a bit of time in the
proposal explaining what it's for and what it does).

al45tair committed:
Update metadata

al45tair committed:
Ensure we have justifications for each API.

Added some text explaining the reason for adding some of the API
surface (some of this was already present).

Also added a section about the previous `Clock` interface with the
`traits` and `convert` functions.

al45tair committed:
Updated after further comments from John.

* Added `ThreadDonationExecutor`.

* Re-worked the other executors protocols to use it.

* Added default implementations for `ExecutorFactory`'s properties.

* Removed extraneous `Optional`s in the task allocator.

* Updated various bits of wording.

al45tair committed:
Updated after further comments on the pitch thread

Added some text about alternatives to `currentExecutor`.  Also
credit John McCall as well, as he's provided substantial input on
the proposal at this point.

al45tair committed:
Remove `Task.currentExecutor` and `Task.schedulingExecutor`.

We've decided to make these non-public.

al45tair committed:
Removed delayed enqueuing from this proposal.

The delayed enqueuing support will be moved to another proposal.

al45tair committed:
Added fourth pitch URL
